packages/block-editor/src/components/inserter/utils.js

```diff
@@ -1,45 +1,71 @@
+// TODO jsdoc + tests
+const getVariationMapper = ( item ) => ( variation ) => ( {
+	...item,
+	id: `${ item.id }-${ variation.name }`,
+	icon: variation.icon || item.icon,
+	title: variation.title || item.title,
+	description: variation.description || item.description,
+	// If `example` is explicitly undefined for the variation, the preview will not be shown.
+	example: variation.hasOwnProperty( 'example' )
+		? variation.example
+		: item.example,
+	initialAttributes: {
+		...item.initialAttributes,
+		...variation.attributes,
+	},
+	innerBlocks: variation.innerBlocks,
+} );
+
 /**
  * Normalizes an inserter block types list and includes variations as separate items.
  *
  * @param {Array} items Denormalized inserter items
+ * @param {number} [limit] Limit returned results by a given threshold. If limit===-1 return all
  * @return {Array} Normalized inserter items.
  */
-export function includeVariationsInInserterItems( items ) {
-	return items.reduce( ( result, item ) => {
-		const { variations = [] } = item;
-		const hasDefaultVariation = variations.some(
-			( { isDefault } ) => isDefault
-		);
+export function includeVariationsInInserterItems( items, limit = -1 ) {
+	const itemsLength = items.length;
+	if ( limit > 0 && itemsLength >= limit ) {
+		// No need to iterate for variations
+		return items.slice( 0, limit );
+	}
+	if ( limit < 1 ) {
+		// Show all available blocks with variations
+		return items.reduce( ( result, item ) => {
+			const { variations = [] } = item;
+			const hasDefaultVariation = variations.some(
+				( { isDefault } ) => isDefault
+			);
 
-		// If there is no default inserter variation provided,
-		// then default block type is displayed.
-		if ( ! hasDefaultVariation ) {
-			result.push( item );
-		}
+			// If there is no default inserter variation provided,
+			// then default block type is displayed.
+			if ( ! hasDefaultVariation ) {
+				result.push( item );
+			}
 
-		if ( variations.length ) {
-			result = result.concat(
-				variations.map( ( variation ) => {
-					return {
-						...item,
-						id: `${ item.id }-${ variation.name }`,
-						icon: variation.icon || item.icon,
-						title: variation.title || item.title,
-						description: variation.description || item.description,
-						// If `example` is explicitly undefined for the variation, the preview will not be shown.
-						example: variation.hasOwnProperty( 'example' )
-							? variation.example
-							: item.example,
-						initialAttributes: {
-							...item.initialAttributes,
-							...variation.attributes,
-						},
-						innerBlocks: variation.innerBlocks,
-					};
-				} )
+			if ( variations.length ) {
+				const variationMapper = getVariationMapper( item );
+				result = result.concat( variations.map( variationMapper ) );
```

Comment:
Given the circumstances, and the fact that we already do it a few lines above, let's replace concat:

```suggestion
				result.push( ...variations.map( variationMapper ) );
```

It [should be safe](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply):

> The consequences of applying a function with too many arguments (that is, **more than tens of thousands** of arguments) varies across engines. (The JavaScriptCore engine has hard-coded argument limit of 65536.

:D